File: app/src/main/java/com/dss886/transmis/base/BaseSwitchActivity.java
package com.dss886.transmis.base;

import android.os.Bundle;
import android.widget.LinearLayout;

import com.dss886.transmis.R;
import com.dss886.transmis.view.BaseItem;
import com.dss886.transmis.view.SectionItem;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.Nullable;

/**
 * Created by duansishu on 2018/1/28.
 */

public abstract class BaseSwitchActivity extends BaseActivity {

    private List<BaseItem> mItems = new ArrayList<>();
    protected LinearLayout mContainer;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mContainer = findViewById(R.id.container);
        addItems();
        setListeners();
        build();
    }

    @Override
    protected void onResume() {
        super.onResume();
        for (BaseItem item : mItems) {
            item.onResume();
        }
    }

    protected void addItem(BaseItem item) {
        mItems.add(item);
    }

    private void build() {
        for (int i = 0; i < mItems.size(); i++) {
            BaseItem item = mItems.get(i);
            if (i == 0 && item instanceof SectionItem) {
                ((SectionItem) item).hideTopPadding();
            }
            if (i > 0 && item instanceof SectionItem) {
                mItems.get(i - 1).hideDivider();
            }
            mContainer.addView(item);
        }
    }

    @Override
    protected int getLayout() {
        return R.layout.activity_base_switch;
    }

    protected abstract void addItems();

    protected abstract void setListeners();
}

File: app/src/main/java/com/dss886/transmis/base/BaseSwitchActivity.kt
package com.dss886.transmis.base

import android.os.Bundle
import android.widget.LinearLayout
import com.dss886.transmis.R
import com.dss886.transmis.utils.buildView
import com.dss886.transmis.utils.forEachChildren
import com.dss886.transmis.view.BaseItemView
import com.dss886.transmis.view.IConfig
import com.dss886.transmis.view.SectionItemView

/**
 * Created by duansishu on 2018/1/28.
 */
abstract class BaseSwitchActivity : BaseActivity() {

    private lateinit var mContainer: LinearLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mContainer = findViewById(R.id.container)
        buildView(getConfigs())
    }

    override fun onResume() {
        super.onResume()
        mContainer.forEachChildren {
            (it as? BaseItemView)?.onResume()
        }
    }

    private fun buildView(configs: List<IConfig>) {
        var lastView : BaseItemView? = null
        configs.forEachIndexed { index, config ->
            val view = config.buildView(this)
            if (index == 0 && view is SectionItemView) {
                view.hideTopPadding()
            }
            if (index > 0 && view is SectionItemView) {
                lastView?.hideDivider()
            }
            mContainer.addView(view)
            lastView = view
        }
    }

    override fun getLayout(): Int {
        return R.layout.activity_base_switch
    }

    protected abstract fun getConfigs(): List<IConfig>

}